src/wallet/rpcwallet.cpp

```diff
@@ -3463,57 +3476,112 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
+            filters.jsop = jsop;
+            filters.pa_sprout = pa;
+
+            if (filters.Common() || filters.Sprout()) {
+                ++itr;
+                continue;
+            }
+
             UniValue obj(UniValue::VOBJ);
-            obj.pushKV("txid", entry.jsop.hash.ToString());
-            obj.pushKV("amount", ValueFromAmount(CAmount(entry.note.value())));
-            obj.pushKV("amountZat", CAmount(entry.note.value()));
-            std::string data(entry.memo.begin(), entry.memo.end());
+
+            obj.pushKV("txid", txid.ToString());
+
+            const auto decrypted = wtx.DecryptSproutNote(jsop).first;
+            obj.pushKV("amount", ValueFromAmount(CAmount(decrypted.note(pa).value())));
+            obj.pushKV("amountZat", CAmount(decrypted.note(pa).value()));
+            obj.pushKV("timestamp", ts);
+            std::string data(decrypted.memo().begin(), decrypted.memo().end());
             obj.pushKV("memo", HexStr(data));
-            obj.pushKV("jsindex", entry.jsop.js);
-            obj.pushKV("jsoutindex", entry.jsop.n);
-            obj.pushKV("confirmations", entry.confirmations);
 
-            txblock BlockData(entry.jsop.hash);
+            obj.pushKV("jsindex", jsop.js);
+            obj.pushKV("jsoutindex", jsop.n);
+            obj.pushKV("confirmations", wtx.GetDepthInMainChain());
+
+            txblock BlockData(txid);
             obj.pushKV("blockheight", BlockData.height);
             obj.pushKV("blockindex", BlockData.index);
             obj.pushKV("blocktime", BlockData.time);
 
             if (hasSpendingKey) {
-                obj.pushKV("change", pwalletMain->IsNoteSproutChange(nullifierSet, entry.address, entry.jsop));
+                obj.push_back(Pair("change", pwalletMain->IsNoteSproutChange(nullifierSet, zaddr, jsop)));
             }
             result.push_back(obj);
+
+            ++itr;
         }
     } else if (boost::get<libzcash::SaplingPaymentAddress>(&zaddr) != nullptr) {
-        for (SaplingNoteEntry & entry : saplingEntries) {
-            UniValue obj(UniValue::VOBJ);
-            obj.pushKV("txid", entry.op.hash.ToString());
-            obj.pushKV("amount", ValueFromAmount(CAmount(entry.note.value())));
-            obj.pushKV("amountZat", CAmount(entry.note.value()));
-            obj.pushKV("memo", HexStr(entry.memo));
-            obj.pushKV("outindex", (int)entry.op.n);
-            obj.pushKV("confirmations", entry.confirmations);
-
-            txblock BlockData(entry.op.hash);
-            obj.pushKV("blockheight", BlockData.height);
-            obj.pushKV("blockindex", BlockData.index);
-            obj.pushKV("blocktime", BlockData.time);
+        const auto pa = *boost::get<libzcash::SaplingPaymentAddress>(&zaddr);
+        const auto itr_sapling = pwalletMain->GetNotesByType(pa.GetHash(), NoteType::sapling, timestamp);
+        auto itr = itr_sapling.begin();
+
+        while (itr != itr_sapling.end() && result.size() < limit) {
```

Comment:
Use a `for` loop as above.

Reply:
https://github.com/zcash/zcash/pull/4399/commits/3138989d2467ac57935bbc14ce0641c6d784f0f2